pyface/color.py

```diff
@@ -127,6 +89,35 @@ class Color(HasStrictTraits):
     #: A tuple holding the hue, lightness, and saturation channels.
     hls = Property(ChannelTuple, depends_on='rgb')
 
+    @classmethod
+    def from_str(cls, text, **traits):
+        """ Create a new Color object from a string.
+
+        Parameters
+        ----------
+        text : str
+            A string holding the representation of the color.  This can be:
+
+            - a color name, including all CSS color names, plus any additional
+              names found in pyface.color.color_table.  The names are
+              normalized to lower case and stripped of whitespace, hyphens and
+              underscores.
+
+            - a hex representation of the color in the form '#RGB', '#RGBA',
+              '#RRGGBB', '#RRGGBBAA', '#RRRRGGGGBBBB', or '#RRRRGGGGBBBBAAAA'.
+
+        **traits
+            Any additional trait values to be passed as keyword arguments.
+
+        Raises
+        ------
+        ColorParseError
+            If the string cannot be converted to a valid color.
+        """
+        space, channels = parse_text(text)
+        traits[space] = channels
```

Comment:
For now, if `traits` already includes `rgba` or `rgb`, it may be silently overridden depending on the value of `space`.

e.g. The `rgb` is overridden here: `Color.from_str("#ccddee", rgb=(1.0, 1.0, 1.0))`.
e.g. But the `rgb` is NOT overridden here because `space` is `"rgba:`: `Color.from_str("#ccddeeff", rgb=(1.0, 1.0, 1.0))`, but the `rgba` value is probably still used because of dictionary ordering.
e.g. The `rgba` is overridden but `rgb` is probably (?) the one ended up being used due to dictionary insertion order being preserved and used by trait instantiation: `Color.from_str("#ccddeeff", rgba=(0.1, 0.2, 0.3, 1.0), rgb=(0.1, 0.2, 0.3))`

While the above examples may not be called directly like that, it could happen programmatically, e.g. when one tries to override traits on the Color instance with a new text:
```
color = Color(rgba=(1.0, 0.2, 0.3, 0.4))
new_color = Color.from_str(user_input, **color.traits())
```
If we are to support non-RGBA with this pattern, this is going to be even more complicated.
Actually, why do we have this `**traits` here? Could we remove it until we need it?

Reply:
For reference:
```
Python 3.8.2 (v3.8.2:7b3ab5921f, Feb 24 2020, 17:52:18) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyface.color import Color
>>> Color.from_str("#aabbccee", rgba=(0.1, 0.2, 0.3, 0.4), rgb=(0.9, 0.9, 0.9))
Color(rgba=(0.9, 0.9, 0.9, 0.9333333333333333))
>>> Color.from_str("#aabbccee", rgb=(0.9, 0.9, 0.9), rgba=(0.1, 0.2, 0.3, 0.4))
Color(rgba=(0.6666666666666666, 0.7333333333333333, 0.8, 0.9333333333333333))
```
(same behaviour on Python 3.6 actually)

Reply:
The `**traits` is defensive programming for subclasses - it's not we who need it, but subclasses that want to call `super()` but need to pass additional traits through.  My general policy is that `@classmethod` factory methods should always:
- call `cls(...)` rather than a hard-coded class name, and
- accept `**traits` and pass them through to the constructor.

There is one immediate case for having `**traits` which is to allow constructs like `Color.from_str('#aabbcc', alpha=0.5)` which seems reasonable to me (not preferred, perhaps, but not disallowed).  This is also a reason to keep the color channel option rather than using `rgba` always.

Everything else seems to fall into a "don't do that" situation (and I can't think of a reason _why_ anyone would try to do something like the examples above).

Reply:
I think I am in general in favour of a simpler API such that it imposes less complexity to the software, so I tend to avoid speculating generality and speculating use cases until we have evidence for it. My approach seems at odds with a lot of what is happening around Color and Font.
@rahulporuri Do you want to chip in your opinion?

Reply:
I think I was hoping that by removing the `**traits`, I would not have to worry about ordering this much. The feature is entirely optional, because one can still do this:
```
color = Color.from_str("#aabbcc")
color.alpha = 0.5
```

If we must keep this, then let's do the ordering right, we should let the `**traits` override the trait parsed from string, not the other way round.

In other words, we should do this ordering instead:
```
kwargs = {space: channel}
kwargs.update(traits)
return cls(**kwargs)
```

The trait name returned by `parse_text` is an _implicit_ implementation detail. Whereas the traits given from the keyword arguments are explicit. It would be surprising to write code like this:
```
color = super().from_str(text, hls=...)
```
and not have the `hls` being set to the requested value because `text` has overruled it.

This would also resolve the weirdness seen in the examples I showed above.
An API is a better API if we don't need to say to developers "don't do that", it should be obvious to the developers what they should do and what they should expect.